### ISSUE TITLE
Anomalies when using `--apply-fixes`

### ISSUE THREAD

**Issue by sebastiaanspeck** [StevenLooman/sonar-magik#196]
I found some anomalies, when using `--apply-fixes`.

```
_proc()
```

Becomes
```
_proc ()
```

And 

```
_loop@label
```

Becomes

```
_loop @label
```

The last one is related to indentation:
```
	>> _if _true
	   _then >> 1
	   _endif.default(2)
```

Becomes
```
	>> _if _true
	_then >> 1
	_endif.default(2)
```

~I do not know if this is desired behaviour.~ If I revert the indentation-changes, the linter triggers on "improper indenting".

```
_pragma(classify_level=basic)
def_slotted_exemplar(:person,
		     {
			     {:name, _unset},
			     {:date_of_birth, _unset}
		     })
$
```

Should be
```
_pragma(classify_level=basic)
def_slotted_exemplar(:person,
	{
		{:name, _unset},
		{:date_of_birth, _unset}
	})
$
```

But I use tree-sitter which resolves the indentation to the first code snippet.

_Originally posted by @sebastiaanspeck in https://github.com/StevenLooman/magik-tools/issues/138#issuecomment-2526100274_

**Comment by sebastiaanspeck** [comment]
The first two examples are fixed with the latest build. Only issue left is the indentation.

**Comment by StevenLooman** [comment]
This is expected as the formatter now drives the check. I plan on making the formatter more configurable and improve the indenting.

**Comment by sebastiaanspeck** [comment]
> I plan on making the formatter more configurable and improve the indenting.

Out of curiosity, how far is the new formatter? I now use a [patched branch](https://github.com/sebastiaanspeck/magik-tools/tree/fix/disable-whitespace) that has indenting turned off

**Comment by StevenLooman** [comment]
There is a formatter (indenter) which is about 95% compatible with the magik-mode indenter. Though the many "features" (or lack thereof) in the magik-mode indenter causes the code to be fairly bloated. I don't want to merge that. For example, the first statement is handled differently (due to the hacked in `point-min` token at the start of the buffer):
```
#_package user  # Commented explicitly, to show this is not a statement

_global glob <<
_proc@glob()
	show(1)
	     _endproc
```

vs

```
_package user

_global glob <<
	_proc@glob()
		show(1)
	_endproc
```

Or missing proper handling of some constructs, such as:
```
_block
	_local val <<
		_for i _over {}
	_loop
		
	_endloop
_endblock
```

There are many more examples. The "new" `_while` construct was also never added.

Also, there is the need to keep track of earlier changes by other parts of the formatter. E.g., when a whitespace is removed some tokens need to moved (read: keep administration for this.) In magik-mode, the buffer is edited inline, but this is not possible the way LSPs work as that returns proposals for edits (`TextEdit`s).

Also, I'm also ignoring the fact that several indenters now exist, outside of `magik-tools`. For example, there are (I'm sure this list is incomplete):
* https://github.com/roadrunner1776/magik/blob/master/magik-indent.el
* https://github.com/roadrunner1776/magik/blob/master/magik-ts-mode.el, for which I cannot find any documentation on how to use this
* The original magik-mode found in the SW(4) distribution
* The vscode extension https://github.com/OpenSmallworld/magik-vscode
* This LSP/formatter
* Several other vscode extensions

Which one to emulate (or all?) is unclear at best to determine at the moment.

All this makes me lose a bit of interest in this. Also, I'm working on other things at the moment, and, of course, life takes time as well.

> Out of curiosity, how far is the new formatter? I now use a [patched branch](https://github.com/sebastiaanspeck/magik-tools/tree/fix/disable-whitespace) that has indenting turned off

You should be able to use the https://github.com/StevenLooman/magik-tools/blob/2b5f3f478c8a4aa21320c2010a656e8f64df3a23/magik-squid/src/main/java/nl/ramsolutions/sw/magik/formatting/NullIndentStrategy.java for this. Though, not configurable at the moment.

**Comment by sebastiaanspeck** [comment]
I understand there are multiple indenters "on the market". You could argue if indentation is a responsibility for the linter itself since indentation is not needed to have working Magik-code, which also makes it easy for tools to write their own indentation rules.

From my knowledge I can say that magik-mode indenter works with elisp-code and regex. The magik-ts-mode, which uses treesit and the [grammar](https://github.com/krn-robin/tree-sitter-magik) to work with nodes to write certain indentation rules.
Since the magik-mode Emacs package added treesit support to Magik, I did not bother the old magik-mode indenter. There are sure limitations which could be fixed, so feel free to open issues there.

I leave it up to you if you really want to have `TextEdit`s for indentation or purely for spaces.

**Comment by StevenLooman** [comment]
> You could argue if indentation is a responsibility for the linter itself since indentation is not needed to have working Magik-code, which also makes it easy for tools to write their own indentation rules.

There are plenty of linters which (can) enforce formatting (which includes indenting.) Furthermore, LSPs can provide [formatting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting) functionality.

> Since the magik-mode Emacs package added treesit support to Magik, I did not bother the old magik-mode indenter. There are sure limitations which could be fixed, so feel free to open issues there.

There are many people still using the GE supplied emacs with the original/GE package. If you would want to enforce formatting (white spacing, indenting) from a pipeline, you have to provide proper support, or ignore the formatting.

Fixing indenting in [emacs-mode](https://github.com/roadrunner1776/magik/) will not change that much, as not everybody is upgrading.

**Comment by sebastiaanspeck** [comment]
> Fixing indenting in [emacs-mode](https://github.com/roadrunner1776/magik/) will not change that much, as not everybody is upgrading.

Nevertheless, there have been two issues fixed in the magik-ts-mode.

Is there a way to have a look at what's coming in the magik-tools? Would be nice if we could slowly work towards a stable 0.11.0 release since it brings a lot of new features and bugfixes.

**Comment by StevenLooman** [comment]
> Would be nice if we could slowly work towards a stable 0.11.0 release since it brings a lot of new features and bugfixes.

Yes, I agree. I was planning to include indenting support compatible with magik-mode and #223, but this will be moved to a future release.

What I do want to include, and am working on currently, is a rewrite of the formatting code to make it better maintainable and extensible in the future. This is needed, as you are the one who said the formatting is currently not acceptable in its current form it I remember correctly. Once the formatting rewrite is done, I'll create a new release.

**Comment by sebastiaanspeck** [comment]
> > Would be nice if we could slowly work towards a stable 0.11.0 release since it brings a lot of new features and bugfixes.
> 
> Yes, I agree. I was planning to include indenting support compatible with magik-mode and #223, but this will be moved to a future release.

Sounds good to me.


> What I do want to include, and am working on currently, is a rewrite of the formatting code to make it better maintainable and extensible in the future. This is needed, as you are the one who said the formatting is currently not acceptable in its current form it I remember correctly. Once the formatting rewrite is done, I'll create a new release.

Indeed. In its current form it is complaining on about almost every line the formatting is not correct.

**Comment by StevenLooman** [comment]
I think this can be closed now.
